appknox/reports.go

```diff
@@ -4,139 +4,67 @@ import (
-		}
-		if res.StatusCode == 404 {
-			return nil, errors.New("File not found")
-		}
-		return nil, err
-	}
-
-	return &resp, nil
+type ReportPreferences struct {
+	ShowAPIScan         bool              `json:"show_api_scan"`
+	ShowManualScan      bool              `json:"show_manual_scan"`
+	ShowStaticScan      bool              `json:"show_static_scan"`
+	ShowDynamicScan     bool              `json:"show_dynamic_scan"`
+	ShowIgnoredAnalyses bool              `json:"show_ignored_analyses_scan"`
+	PCIDSSPreferences   PCIDSSPreferences `json:"show_hipaa"`
+	HIPAAPreferences    HIPAAPreferences  `json:"show_pcidss"`
 }
 
-// FetchReportResult it will fetch report result by result id.
-func (s *ReportsService) FetchReportResult(ctx context.Context, reportID int) (*ReportResult, error) {
-	url := fmt.Sprintf("api/v2/reports/%d", reportID)
-	req, err := s.client.NewRequest("GET", url, nil)
-	if err != nil {
-		return nil, err
-	}
-
-	var resp ReportResult
-
-	_, err = s.client.Do(ctx, req, &resp)
-	if err != nil {
-		return nil, err
-	}
+type ReportResult struct {
+	ID                int               `json:"id"`
+	GeneratedOn       *time.Time        `json:"generated_on"`
+	Language          string            `json:"language"`
+	Progress          int               `json:"progress"`
+	Rating            string            `json:"rating"`
+	ReportPreferences ReportPreferences `json:"preferences"`
+}
 
-	return &resp, nil
+type DRFResponseReport struct {
+	Count    int             `json:"count,omitempty"`
+	Next     string          `json:"next,omitempty"`
+	Previous string          `json:"previous,omitempty"`
+	Results  []*ReportResult `json:"results"`
 }
 
-// FetchLastReportResult it will return last report result, report list api is responding in decending order.
-func (s *ReportsService) FetchLastReportResult(ctx context.Context, fileID int) (*ReportResult, error) {
+func (s *ReportsService) List(ctx context.Context, fileID int) ([]*ReportResult, error) {
 	url := fmt.Sprintf("api/v2/files/%d/reports", fileID)
-	req, err := s.client.NewRequest("GET", url, nil)
+	request, err := s.client.NewRequest("GET", url, nil)
 	if err != nil {
 		return nil, err
 	}
 
-	type ReportResultList struct {
-		Results []*ReportResult `json:"results"`
-	}
-
-	var resp ReportResultList
-
-	_, err = s.client.Do(ctx, req, &resp)
-	if err != nil {
-		return nil, err
-	}
+	var drfResponseReport DRFResponseReport
 
-	if len(resp.Results) > 0 {
-		return resp.Results[0], nil
-	}
-	return nil, errors.New("No report results found")
-}
-
-// GetReportURL returns the url of the report file to download.
-func (s *ReportsService) GetReportURL(ctx context.Context, reportID int) (*Report, error) {
-	u := fmt.Sprintf("api/v2/reports/%d/pdf", reportID)
-	req, err := s.client.NewRequest("GET", u, nil)
-	if err != nil {
-		return nil, err
+	resp, err := s.client.Do(ctx, request, &drfResponseReport)
+	if resp.StatusCode == 404 {
```

Comment:
Handle `err` part first and then check for `resp.StatusCode`. Also the status code here can be 403 as well. You can handle 403 separately or provide a generic error for both status codes.

Reply:
Handling err part first will lead to returning generic error and status code error will never be returned.
I have added null pointer check for resp and statusCode check

Reply:
403 check will be done in a separate PR. It will be mostly generic irrespective of command, So planning to change it in a separate PR